app/src/main/java/org/mozilla/focus/browser/BrowserNavigationOverlay.kt

```diff
@@ -151,10 +161,28 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
             else -> R.id.navButtonForward
         }
 
-        navUrlInput.setText(navigationStateProvider?.getCurrentUrl())
+        maybeUpdateOverlayURLForCurrentState()
 
         if (findFocus() == null) {
             requestFocus()
         }
     }
+
+    private fun maybeUpdateOverlayURLForCurrentState() {
+        // The url can get updated in the background, e.g. if a loading page is redirected. We
+        // don't want a url update to interrupt the user typing so we don't update the url from
+        // the background if the user has already updated the url themselves.
+        //
+        // We revert this state when the view is unfocused: it ensures the URL is usually accurate
+        // (for security reasons) and it's simple compared to other options which keep more state.
+        //
+        // One problem this solution has is that if the URL is updated in the background rapidly,
+        // sometimes key events will be dropped, but I don't think there's much we can do about this:
+        // we can't determine if the keyboard is up or not and focus isn't a good indicator because
```

Comment:
> if a user is using the app to enter text, I assume the keyboard wouldn't even pop up

Actually, it generally does (good call, I forgot about the remote app).

> because they can enter directly into an EditText field

They can if they press the keyboard button to manually raise the keyboard, but there's actually a bug open for this not working #566.

> so if the user opens the keyboard and hasn't "entered" anything yet (e.g., moving around the keyboard), this string would still be changing, right? Is there another place to set this? What about in onKeyPreIme?

Yes, the string will change if they're moving around the keyboard. However, there is not another place that I'm aware of that we can call this. `onKeyPreIme` is a good idea, but it's not called for the remote app so we can't use it to fix this problem for the remote app.